File: Design-Patterns-Using-Swift/Factory/FactoryMethodAnimal.swift
// Animal protocol
protocol Animal {
    func voice()
}

// Concrete Animal classes
class Dog: Animal {
    func voice() {
        print("Bhow Bhow!!")
    }
}

class Cat: Animal {
    func voice() {
        print("Meow Meow!!")
    }
}

// AnimalFactory protocol
protocol AnimalFactory {
    func getAnimal() -> Animal
}

// Concrete AnimalFactory classes
class CatFactory: AnimalFactory {
    func getAnimal() -> Animal {
        return Cat()
    }
}

class DogFactory: AnimalFactory {
    func getAnimal() -> Animal {
        return Dog()
    }
}

// Client code
let dogFactory: AnimalFactory = DogFactory()
dogFactory.getAnimal().voice()

let catFactory: AnimalFactory = CatFactory()
catFactory.getAnimal().voice()

// Future changes to include cow type of objects.
class Cow: Animal {
    func voice() {
        print("Gooaa Gooaa!!")
    }
}

class CowFactory: AnimalFactory {
    func getAnimal() -> Animal {
        return Cow()
   }
}

// Client code for Cow
let cowFactory: AnimalFactory = CowFactory()
cowFactory.getAnimal().voice()

/*
Bhow Bhow!!
Meow Meow!!
Gooaa Gooaa!!
*/

File: Design-Patterns-Using-Ruby/Factory/FactoryMethodAnimal.rb
# Animal interface
class Animal
    def voice
        raise NotImplementedError, 'Subclasses must implement the voice method'
    end
end

# Concrete Animal classes
class Dog < Animal
    def voice
        puts 'Bhow Bhow!!'
    end
end

class Cat < Animal
    def voice
        puts 'Meow Meow!!'
    end
end

# AnimalFactory interface
class AnimalFactory
    def get_animal
        raise NotImplementedError, 'Subclasses must implement the get_animal method'
    end
end

# Concrete AnimalFactory classes
class CatFactory < AnimalFactory
    def get_animal
        Cat.new
    end
end

class DogFactory < AnimalFactory
    def get_animal
        Dog.new
    end
end

# Client code
dog_factory = DogFactory.new
dog_factory.get_animal.voice

cat_factory = CatFactory.new
cat_factory.get_animal.voice

# Future changes to include cow type of objects.
class Cow < Animal
    def voice
        puts 'Gooaa Gooaa!!'
    end
end

class CowFactory < AnimalFactory
    def get_animal
        Cow.new
    end
end

# Client code for Cow
cow_factory = CowFactory.new
cow_factory.get_animal.voice

=begin 
Bhow Bhow!!
Meow Meow!!
Gooaa Gooaa!!
=end